x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -125,6 +127,7 @@ protected static Batch<LogicalPlan> substitutions() {
             // lastly replace surrogate functions
             new SubstituteSurrogates(),
             new ReplaceRegexMatch(),
+            new ReplaceTypeConvertingEvalWithAliasingEval(),
```

Comment:
Maybe one thing: This is a proper optimization, not a substitution. I'd expect this to be placed in the `operators` batch, among the `prune/elimination` optimization rules.

Reply:
Essentially my rule needs to occur after `ReplaceStatsNestedExpressionWithEval` and before `ReplaceAliasingEvalWithProject` in order to have the desired effect.